### PR TITLE
collection of BtUTC bugfixes

### PR DESCRIPTION
**Pull request by jebeck**
**WIP. Don't merge.**

At minimum, this PR will contain:

- [x] fix for off-by-hour issue (raised by several Dexcom users but not actually exclusively to Dexcom)
- [x] increase in threshold for rounding timezone offsets (to nearest 30 instead of nearest 15 minutes)

It may also include:

- [x] making BtUTC robust to change for DST _ahead_ of actual change
- [ ] fix for OmniPod PDM reset issue causing some crucial time changes to be dropped

Closes #237, which was stale. (I rebased it onto master here.)

Some notes on the fixes:

- Both the off-by-hour issue and the making BtUTC robust to changing the device time *ahead* of a DST switchover were solved by making the same change: making sure that we use the `mostRecent` timestamp passed to the `TimezoneOffsetUtil` constructor to determine the initial `timezoneOffset` from which to start bootstrapping. Now, the documentation has always said that we *do* determine the initial `timezoneOffset` from the `mostRecent` timestamp, but in fact we were not previously (😧!); instead, we were finding the initial `timezoneOffset` from the most recent `timeChange` in the data (an oversight that slipped by due to some sloppy test fixtures (😱!)). Most of the time this works, because most of the time the most recent time change is a change to the same `timezoneOffset` as `mostRecent`.
- The only other change here (so far) is just to increase the rounding of `timezoneOffset` to the nearest 30 rather than the nearest 15 minutes because we discovered users who do have clock drift adjustments larger than the previous threshold. This change in the code is pretty arbitrary and low-risk; tests and documentation have been updated and that should be reviewed, but the logic doesn't need careful thinking through (as opposed to the previous change).

### CODE REVIEW

**Comment by jebeck** [comment]
<a href="https:&#x2F;&#x2F;trello.com&#x2F;c&#x2F;SyEwM2Jt&#x2F;9-0-ux-2-dev-1-tst-uploader-dexcom-bug-of-one-hour-off"> &lt;0:ux 2:dev 1:tst&gt; [uploader] [dexcom] Bug of one hour off</a>

**Comment by jebeck** [comment]
<a href="https:&#x2F;&#x2F;trello.com&#x2F;c&#x2F;vXyhtzSI&#x2F;17-uploader-ingestion-processing-of-clock-drift-of-7-5-minutes-causes-duplicate-offset-data"> [uploader] [ingestion] processing of clock drift of &gt; 7.5 minutes causes duplicate offset data</a>

**Comment by jebeck** [comment]
<a href="https:&#x2F;&#x2F;trello.com&#x2F;c&#x2F;Gj7xbiIw&#x2F;18-uploader-make-btutc-robust-to-changing-for-dst-ahead-of-actual-switchover"> [uploader] make BtUTC robust to changing for DST *ahead* of actual switchover</a>

File: lib/TimezoneOffsetUtil.js [hunk 1]
```diff
@@ -93,8 +98,12 @@ module.exports = function(timezone, mostRecent, changes) {
     for (var i = 0; i < sorted.length; ++i) {
       var change = sorted[i];
       if (i === 0) {
-        change.time = sundial.applyTimezoneAndConversionOffset(change.jsDate, timezone, conversionOffset).toISOString();
-        change.timezoneOffset = sundial.getOffsetFromZone(change.time, timezone);
```
Comment (jebeck) [inline lib/TimezoneOffsetUtil.js]: The old line 97 is the offending one, replaced by new line 106, using `mostRecent` as the input to sundial's `getOffsetFromZone`.

**Comment by jebeck** [comment]
<a href="https:&#x2F;&#x2F;trello.com&#x2F;c&#x2F;0rOqz1cv&#x2F;22-uploader-review-build-collection-of-btutc-fixes-for-testing"> [uploader] review &amp; build collection of BtUTC fixes for testing</a>

**Comment by gniezen** [comment]
LGTM!

**Comment by gniezen** [comment]
Merging.

**Pull request merged**